x/staking/keeper/unbonding.go

```diff
@@ -0,0 +1,419 @@
+package keeper
+
+import (
+	"encoding/binary"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// Increments and returns a unique ID for an unbonding operation
+func (k Keeper) IncrementUnbondingId(ctx sdk.Context) (unbondingId uint64) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.UnbondingIdKey)
+
+	if bz == nil {
+		unbondingId = 0
+	} else {
+		unbondingId = binary.BigEndian.Uint64(bz)
+	}
+
+	unbondingId = unbondingId + 1
+
+	// Convert back into bytes for storage
+	bz = make([]byte, 8)
+	binary.BigEndian.PutUint64(bz, unbondingId)
+
+	store.Set(types.UnbondingIdKey, bz)
+
+	return unbondingId
+}
+
+// Remove a mapping from UnbondingId to unbonding operation
+func (k Keeper) DeleteUnbondingIndex(ctx sdk.Context, id uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.GetUnbondingIndexKey(id))
+}
+
+// return a unbonding delegation that has an unbonding delegation entry with a certain ID
+func (k Keeper) GetUnbondingDelegationByUnbondingId(
+	ctx sdk.Context, id uint64,
+) (ubd types.UnbondingDelegation, found bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	ubdeKey := store.Get(types.GetUnbondingIndexKey(id))
+	if ubdeKey == nil {
+		return types.UnbondingDelegation{}, false
+	}
+
+	value := store.Get(ubdeKey)
+	if value == nil {
+		return types.UnbondingDelegation{}, false
+	}
+
+	ubd, err := types.UnmarshalUBD(k.cdc, value)
+	// An error here means that what we got wasn't the right type
+	if err != nil {
+		return types.UnbondingDelegation{}, false
+	}
+
+	return ubd, true
+}
+
+// return a unbonding delegation that has an unbonding delegation entry with a certain ID
+func (k Keeper) GetRedelegationByUnbondingId(
+	ctx sdk.Context, id uint64,
+) (red types.Redelegation, found bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	redKey := store.Get(types.GetUnbondingIndexKey(id))
+	if redKey == nil {
+		return types.Redelegation{}, false
+	}
+
+	value := store.Get(redKey)
+	if value == nil {
+		return types.Redelegation{}, false
+	}
+
+	red, err := types.UnmarshalRED(k.cdc, value)
+	// An error here means that what we got wasn't the right type
+	if err != nil {
+		return types.Redelegation{}, false
+	}
+
+	return red, true
+}
+
+// return the validator that is unbonding with a certain unbonding op ID
+func (k Keeper) GetValidatorByUnbondingId(
+	ctx sdk.Context, id uint64,
+) (val types.Validator, found bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	valKey := store.Get(types.GetUnbondingIndexKey(id))
```

Comment:
Exactly. The UnbondingIndexKey is a mapping from UnbondingIDs to strings, where each string is the key of an unbonding operation, i.e., validator (that is unbonding), undelegation entry, redelegation entry.

Reply:
Interesting. This initially seems like a codesmell/poor design choice to me. Is it possible that a value overwrite an existing ID's value unintentionally?

Reply:
You use a dedicated key for each stored value type. Having to type check/error check based on the value is a codesmell. I.e. "I'm storing value X at key K. What is X? Idk, let me try these 5 different types..."

Reply:
generally agree here. This is not a design we want to promote in the sdk

Reply:
Suggestion of how else to do this? Basically, we want to match an ID (unbonding ID) with a key that can have one of the following formats:
- `0x21 | OperatorAddrLen (1 byte) | OperatorAddr`
- `0x32 | DelegatorAddrLen (1 byte) | DelegatorAddr | ValidatorAddrLen (1 byte) | ValidatorAddr`
- `0x34 | DelegatorAddrLen (1 byte) | DelegatorAddr | ValidatorAddrLen (1 byte) | ValidatorSrcAddr | ValidatorDstAddr`

I guess we could change the type of the unbonding ID from `uint64` to `[]byte` and add a prefix depending on the type of unbonding (e.g., `0x44 | UnbondingIndex` for validator unbondings). What do you think? 

@jtremback ^

Reply:
This needs to be implemented though and to be made consistent with the Interchain Security implementation.

Reply:
I don't intend to have this be blocking, I'm just pointing out my opinion that it's a codesmell. I'll leave it to the IS team for how to proceed.

Happy to approve the PR nevertheless 👍

Reply:
Got a second opinion form @ethanfrey and he tends to agree this approach is a code smell. We should refactor this. Would consider this a blocker for the merge. @mpoke any chance you or someone on your team can pick this up. once completed we can merge this work.

Reply:
Lol so you guys just want us to copy and paste it a few times? I can do that for you

Reply:
@marbar3778 @alexanderbez I will take a look at this right now. I wrote the offending code

Reply:
The pattern in question is most clearly demonstrated not here, but in the top level code that calls this function in `UnbondingCanComplete`. `UnbondingCanComplete` works sort of like an if/else statement.

```
if (is redelegation) {

} else if (is undelegation) {

} else if (is validator unbonding) {

} else {

}
```

This didn't seem so bad when I wrote it, but I guess the part that troubles you is just the fact that references to records of different types are being kept in one store? Or is it that we are checking which type it is with failed deserializations?

The simplest way I can think of to avoid the latter issue is to make another index which stores `unbondingId->type`. Then we would check this first to find out what to deserialize to without the potential several failed deserializations which are bugging you.

Reply:
> but I guess the part that troubles you is just the fact that references to records of different types are being kept in one store?

Yes, exactly! It's that a key, a dedicated prefix really, can store values of _different_ types yielding the need to have custom business logic of these if/else blocks as you pointed out. To me this is a code smell and no other SDK module does this.

What if we just combined/flattened all the possible value types into one type and have a field/enum that describes what kind of action to take?

something like

```go
type MyValue struct {
  Type string // or enum

 // all fields that can represent everything you need
}
```

Reply:
Seems like maybe this is something a proto oneof could deal with efficiently